Content/posts/otel-integration.md

```diff
@@ -0,0 +1,298 @@
+
+**Instrumenting** a system means adding observability capabilities. The idea is the code emits the data which then must be collected and sent to a backend observability system. In this blog post we'll focus on metrics, but the same principles apply to logs and traces as well.
+
+### Gathering metrics
+
+The first step is gathering the metrics from our app's flow. Fortunately for us this is the easiest step, as, just like logs with swift-log, Vapor comes with built-in support for metrics using the [swift-metrics](https://github.com/apple/swift-metrics) package.
+
+The metrics created by Vapor include things like request counts, response times, and error rates, all of which are essential for understanding the health and performance of your application. 
+
+> You can also emit your own custom metrics using the same package. To do this, check out the [Swift Metrics documentation](https://swiftpackageindex.com/apple/swift-metrics/2.7.1/documentation/coremetrics). Since Vapor already embeds it, you don't need to configure anything, just add your own metrics where you need them and they will be emitted along with the built-in ones.
+
+### Emitting Metrics
+
+The next step is sending the emitted metrics to a service that can process and store them. To do that we'll be using something known as a sidecar, which is a separate process that runs alongside your application and collects the metrics emitted by it. In our case, we'll be using the [OpenTelemetry Collector](https://opentelemetry.io/). The OTel Collector is a vendor-agnostic tool to receive, process, and export telemetry data. It can receive data in various formats, process it, and export it to various backends.
+
+To interact with the OTel Collector we'll need a client that speaks the same language, which is the OpenTelemetry protocol (OTLP). Luckily, at the time of writing, the [swift-otel](https://github.com/swift-otel/swift-otel) package has just hit version 1.0 and is right up our alley.
+
+Let's add the package to our dependencies in the `Package.swift` file:
+
+```swift
+.package(url: "https://github.com/swift-otel/swift-otel.git", from: "1.0.0"),
+```
+
+and then add it to our target dependencies:
+
+```swift
+.target(
+    name: "App",
+    dependencies: [
+        .product(name: "OTel", package: "swift-otel"),
+    ]
+),
+```
+
+
+While the package is designed to work with swift-service-lifecycle which is not Vapor 4's favourite lunch buddy, we can integrate it manually into our application's lifecycle. To do this we can create a `LifecycleHandler` that will start and stop the OTel metrics exporter when the application starts and stops.
+
+```swift
+import OTel
+import ServiceLifecycle
+import Vapor
+
+actor OTelMetricsExporterLifecycleHandler: LifecycleHandler {
```

Comment:
I understand you're just bootstrapping Metrics in this post but I could see how users copy-paste this for their use cases with Logs, Metrics, and Tracing. How about calling this `OTelLifecycleHandler` instead?